File: ios/App/App/Transport/ReconnectBackoff.swift
import Foundation

/// Full-jitter exponential reconnect backoff matching `src/shared/remote/backoff.ts`.
struct ReconnectBackoff: Sendable {
    private(set) var attempt: Int = 0
    var baseMs: Double
    var maxMs: Double

    init(
        baseMs: Double = RemoteSocketPolicy.reconnectBaseMs,
        maxMs: Double = RemoteSocketPolicy.reconnectMaxMs
    ) {
        self.baseMs = baseMs
        self.maxMs = maxMs
    }

    mutating func nextDelay() -> Duration {
        let ceiling = min(maxMs, baseMs * pow(2.0, Double(attempt)))
        let delayMs = ceiling / 2 + Double.random(in: 0 ..< max(ceiling / 2, 0.000_001))
        attempt += 1
        return .milliseconds(Int64(delayMs.rounded()))
    }

    mutating func reset() {
        attempt = 0
    }
}

File: android/app/src/main/kotlin/com/poracode/app/protocol/ReconnectBackoff.kt
package com.poracode.app.protocol

import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.random.Random

/**
 * Full-jitter exponential reconnect backoff matching `src/shared/remote/backoff.ts`:
 * `min(maxMs, baseMs * 2**attempt)` then jitter into `[ceiling/2, ceiling)`.
 */
class ReconnectBackoff(
    private val baseMs: Long = RemoteSocketPolicy.RECONNECT_BASE_MS,
    private val maxMs: Long = RemoteSocketPolicy.RECONNECT_MAX_MS,
    private val random: Random = Random.Default,
) {
    var attempt: Int = 0
        private set

    fun nextDelayMs(): Long {
        val ceiling = min(maxMs.toDouble(), baseMs * 2.0.pow(attempt.toDouble()))
        val half = ceiling / 2.0
        val delay = half + random.nextDouble() * max(half, 0.000_001)
        attempt += 1
        return delay.toLong()
    }

    fun reset() {
        attempt = 0
    }
}
